lighthouse-core/audits/byte-efficiency/unused-javascript.js

```diff
@@ -65,43 +71,148 @@ class UnusedJavaScript extends ByteEfficiencyAudit {
     }
 
     return {
+      unusedByIndex,
       unusedLength: unused,
       contentLength: maximumEndOffset,
     };
   }
 
   /**
-   * @param {Array<{unusedLength: number, contentLength: number}>} wasteData
-   * @param {LH.Artifacts.NetworkRequest} networkRecord
+   * @param {LH.Audit.ByteEfficiencyItem} item
+   * @param {WasteData[]} wasteData
+   * @param {import('../../computed/bundle-analysis.js').Bundle} bundle
+   * @param {ReturnType<typeof UnusedJavaScript.determineLengths>} lengths
+   */
+  static createBundleMultiData(item, wasteData, bundle, lengths) {
+    if (!bundle.script.content) return;
+
+    /** @type {Record<string, number>} */
+    const files = {};
+
+    // This is 10x slower (~320ms vs 46ms for a big map), but its correctness
+    // is much easier to reason about. The latter method gives the same counts,
+    // except it seems to have +1 byte for each file.
+    // let line = 0;
+    // let column = 0;
+    // for (let i = 0; i < bundle.script.content.length; i++) {
+    //   column += 1;
+    //   if (bundle.script.content[i] === '\n') {
+    //     line += 1;
+    //     column = 0;
+    //   }
+    //   if (wasteData.some(data => data.unusedByIndex[i] === 0)) continue;
+
+    //   // @ts-ignore: ughhhhh the tsc doesn't work for the compiled cdt lib
+    //   const mapping = bundle.map.findExactEntry(line, column);
+    //   // This can be null if the source map has gaps.
+    //   // For example, the webpack CommonsChunkPlugin emits code that is not mapped (`webpackJsonp`).
+    //   if (mapping) {
+    //     files[mapping.sourceURL] = (files[mapping.sourceURL] || 0) + 1;
+    //   }
+    // }
+
+    const lineLengths = bundle.script.content.split('\n').map(l => l.length);
+    let totalSoFar = 0;
+    const lineOffsets = lineLengths.map(len => {
+      const retVal = totalSoFar;
+      totalSoFar += len + 1;
+      return retVal;
+    });
+
+    // @ts-ignore
+    bundle.map.computeLastGeneratedColumns();
+    // @ts-ignore
+    for (const mapping of bundle.map.mappings()) {
+      let offset = lineOffsets[mapping.lineNumber];
+
+      offset += mapping.columnNumber;
+      const byteEnd = mapping.lastColumnNumber || lineLengths[mapping.lineNumber];
+      for (let i = mapping.columnNumber; i < byteEnd; i++) {
+        if (wasteData.every(data => data.unusedByIndex[offset] === 1)) {
```

Comment:
i guess the alternative to this is if the mapping could also express its start and endOffsets?

i'm interested in what'd be involved in making that happen.

Reply:
It would be straight-forward, but would need to be done upstream in CDT. The offset can be pulled from the character iterator's `position` (see `SourceMap._parseMap`).

This current method is good for now. when you bring this feature to CDT it'll be good if you can make the unused calculation re-usable, then we can delete this function.

Reply:
> This current method is good for now. when you bring this feature to CDT it'll be good if you can make the unused calculation re-usable, then we can delete this function.

lol so you're saying that what you did is fine, but i should reimplement when i upstream it? 🤣 


i guess you're taking this position mostly because the changes to sdk/sourcemap.js would be too significant? i guess that makes sense, though the class now is pretty modified already.   ah well.

Reply:
There's two parts here:

1. modify `SourceMap._parseMap` to also save the end offsets, so the byte attribution can be a little faster
2. the feature you're working on in CDT will need to do the same byte attribution. So there is no difference in LH vs CDT, this should be done in the same way by using the same code

I don't want to do 1. here because the change needed is much more than how this PR is currently augmenting SourceMap/SDK. Would need to completely replace a private function. Since it's just for speed, not correctness, I think it can wait until done upstream.

For 2, I'd imagine you'll just copy L90-116 here, and make a function on SourceMap, something like `calculateSourceByteAttribution(compiledSource: string)`. I'm lost on if you've done this bit already in your CL - are you already calculating attribution for bundle?

Reply:
also, I don't mind taking over the above within DT :)

Reply:
> also, I don't mind taking over the above within DT :)

that sounds pretty divine to me.